### ISSUE TITLE
Documentation on appRoleAssignment API is incorrect

### ISSUE THREAD

**Issue by Neutrino-Sunset** [microsoftgraph/microsoft-graph-docs#22207]
The documentation [here](https://learn.microsoft.com/en-us/graph/api/serviceprincipal-post-approleassignments?view=graph-rest-1.0&tabs=http#permissions) states that one of these roles is required to call the appRoleAssignment api.

> AppRoleAssignment.ReadWrite.All and Application.Read.All, AppRoleAssignment.ReadWrite.All and Directory.Read.All

So how come a service principal having every single one of these roles still fails with `Insufficient privileges to complete the operation`?

![image](https://github.com/microsoftgraph/microsoft-graph-docs/assets/18445931/3dc916db-b2c8-451a-b12c-b0a273c25d29)

![image](https://github.com/microsoftgraph/microsoft-graph-docs/assets/18445931/eb43d583-eebf-4c81-ab54-938ebda0c9e6)

![image](https://github.com/microsoftgraph/microsoft-graph-docs/assets/18445931/ede7439d-f034-40ff-90c1-c4d15f9c0175)

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 47f3b787-b279-695c-8681-7832b203d5b8
* Version Independent ID: 794ffd87-99dd-0521-cd5d-8a340d5babad
* Content: [Grant an appRoleAssignment to a service principal - Microsoft Graph v1.0](https://learn.microsoft.com/en-us/graph/api/serviceprincipal-post-approleassignments?view=graph-rest-1.0&tabs=http)
* Content Source: [api-reference/v1.0/api/serviceprincipal-post-approleassignments.md](https://github.com/microsoftgraph/microsoft-graph-docs/blob/main/api-reference/v1.0/api/serviceprincipal-post-approleassignments.md)
* Product: **applications**
* Technology: **microsoft-graph**
* GitHub Login: @psignoret
* Microsoft Alias: **MSGraphDocsVteam**

**Comment by Neutrino-Sunset** [comment]
So I've got this working.

It turns out that in addition to the Graph API roles described here, any service principal calling the `appRoleAssignments` API also requires the Azure AD role `Privileged Role Administrator`.

Neither the documentation nor the error message mentions this, so I've no idea how anyone is supposed to work this out.

Graph API is promoted as being a replacement for the Azure AD API, yet doing anything with the Graph API requires having a combination of Graph API permissions and Azure AD roles which does not seem to be documented anywhere. This is really confusing. I would have thought it would make more sense if the Graph API permissions automatically applied the necessary Azure AD roles, otherwise you just end up having to use multiple APIs or a combination of the Graph API and the Azure Portal in order to get done something that you should be able to get done using the Graph API alone.

To put it another way, if you can't get done what you need to using the Graph API alone, then it's not a very useful or user friendly API.

**Comment by Neutrino-Sunset** [comment]
This is what the Azure Portal page showed when I made the request.

![image](https://github.com/microsoftgraph/microsoft-graph-docs/assets/18445931/22c33e50-5743-4b77-bebd-6fdf098f8d4f)

**Comment by Neutrino-Sunset** [comment]
I've tested it again, and I have to admit you seem to be correct.

I can only assume that after adding the `AppRoleAssignment.ReadWrite.All` and `Application.ReadWrite.All` there was a period of time where those roles were not taking effect during my testing.

**Comment by psignoret** [comment]
@Neutrino-Sunset The token used by `az rest` was a cached token that had been obtained prior to when the permission was granted. Running `az login` only gets a fresh token to Azure Management, but other cached tokens, such as tokens to Microsoft Graph, are still there and will be used if they haven't expired. To do a clean test, you need to use `az account clear` to remove all cached tokens:

1. Remove all directory role assignments from the service principal.
2. Verify that the desired API permissions are granted.
3. If you made any changes to directory role assignments or API permission grants, go grab a fresh cup of coffee (i.e. wait a few minutes).
4. Run `az account clear` to clear all signed-in accounts in Azure CLI.
5. Run `az login --service-principal ...` to authenticate as the service principal.
6. Run `az rest --method post ...` to make the API request.
